test/memory/memory_api_test.go

```diff
@@ -0,0 +1,260 @@
+	metricsBefore.readMetrics(samples)
+	for i := 0; i < int(repeats); i++ {
+		f()
+	}
+	m.readMetrics(samples)
+	m.d = time.Since(now)
+	m.rps = int(float64(repeats) / m.d.Seconds())
+
+	defer m.print()
+	return m.diff(metricsBefore)
+}
+
+// memMetrics is a list of metrics relevant to the memory test. Metrics suffixed with 'Max' can be pre-defined
+// to serve as test criteria (these are not filled during 'readMetrics()').
+type memMetrics struct {
+	totalAlloc      uint64
+	totalAllocMax   uint64
+	memoryFreed     uint64
+	heapAlloc       uint64
+	heapAllocMax    uint64
+	objectAlloc     uint64
+	objectFreed     uint64
+	objectRemain    uint64
+	objectRemainMax uint64
+	goroutines      int
+	rps             int
+	d               time.Duration
+}
+
+// readMetrics populates memory metrics
+func (m *memMetrics) readMetrics(s []metrics.Sample) {
+	metrics.Read(s)
+	for _, sample := range s {
+		switch sample.Name {
+		case heapBytesAllocs:
+			m.totalAlloc = sample.Value.Uint64()
+		case heapBytesFrees:
+			m.memoryFreed = sample.Value.Uint64()
+		case heapObjectAllocs:
+			m.objectAlloc = sample.Value.Uint64()
+		case heapObjectFrees:
+			m.objectFreed = sample.Value.Uint64()
+		case heapObjects:
+			m.objectRemain = sample.Value.Uint64()
+		case goroutines:
+			m.goroutines = int(sample.Value.Uint64())
+		}
+	}
+	m.heapAlloc = m.totalAlloc - m.memoryFreed
+}
+
+// compares metrics with another metric snapshot taken earlier. Calculates difference between those
+// and for selected metrics evaluates pass/fail criteria
+func (m *memMetrics) diff(before *memMetrics) (pass bool) {
+	pass = true
+	m.totalAlloc -= before.totalAlloc
+	if m.totalAlloc > m.totalAllocMax {
+		pass = false
+	}
+	m.memoryFreed -= before.memoryFreed
+	m.heapAlloc -= before.heapAlloc
+	if m.heapAlloc > m.heapAllocMax {
+		pass = false
+	}
+	m.objectAlloc -= before.objectAlloc
+	m.objectFreed -= before.objectFreed
+	m.objectRemain -= before.objectRemain
+	if m.objectRemain > m.objectRemainMax {
+		pass = false
+	}
+	return
+}
+
+func (m *memMetrics) print() {
+	p := message.NewPrinter(language.English)
+	var totalAllocMax, heapAllocMax, objectReaminMax string
+	// if thresholds are set, use them in the output
+	if !(m.totalAllocMax == 0 && m.heapAllocMax == 0 && m.objectRemainMax == 0) {
+		totalAllocMax = format(m.totalAllocMax)
+		heapAllocMax = format(m.heapAllocMax)
+		objectReaminMax = p.Sprintf("%d", m.objectRemainMax)
+	}
+	fmt.Printf("Total alloc:\t%s\t%s\n", format(m.totalAlloc), totalAllocMax)
+	fmt.Printf("Memory Freed:\t%s\n", format(m.objectFreed))
+	fmt.Printf("Heap alloc:\t%s\t%s\n", format(m.heapAlloc), heapAllocMax)
+	fmt.Printf("Objects alloc:\t%s\nObj freed:\t%s\n", p.Sprintf("%d", m.objectAlloc),
+		p.Sprintf("%d", m.objectFreed))
+	fmt.Printf("Objects remain:\t%s\t\t%s\n", p.Sprintf("%d", m.objectRemain), objectReaminMax)
+	fmt.Printf("Message rate:\t%d m/s\n", m.rps)
+	fmt.Printf("Num goroutines:\t%d\nDuration:\t%s\n\n", m.goroutines, m.d.String())
+}
```

Comment:
You could make it even easier to add new metrics by defining a map here:
```suggestion
type metric struct {
	value uint64
	max *uint64
}

type memMetrics struct {
	metricByName map[string]metric
	d time.Duration
}

// readMetrics populates memory metrics
func (m *memMetrics) readMetrics(s []metrics.Sample) {
	metrics.Read(s)
	for _, sample := range s {
		m.metricByName[sample.Name].value = sample.Value.Uint64()
	}
}

func (m *memMetrics) diff(before *memMetrics) (pass bool) {
	pass = true
	for name, metric := range m.metricByName {
		metric.value -= before.metricByName[name].value
		if metric.max != nil && metric.value > metric.max {
			pass = false
		}
	}
}

func (m *memMetrics) print() {
	...
	template := "%s:\t%s\n"
	templateWithMax := "%s:\t%s\t%s\n"
	for name, metric := range m.metricByName {
		if metric.max != nil {
			fmt.Printf(templateWithMax, name, format(metric.value), metric.max)
		} else {
			fmt.Printf(template, name, format(metric.value))
		}
	}
	...
}
```

Reply:
Done, thanks for the suggestion.